### ISSUE TITLE
XLA optimisations can discard stack trace metadata

### ISSUE THREAD

**Issue by olupton** [openxla/xla#9542]
In a simple JAX program, such as
```python
import jax
import jax.numpy as jnp

def bar(z):
  return z + 1

def glug(z):
  return z + 2

@jax.jit
def prog(x):
    y = glug(x)
    z = bar(y)
    return z

x = jnp.array([1,2,3])
prog(x)
```
which yields the jaxpr
```
{ lambda ; a:i32[3]. let
    b:i32[3] = add a 2   # /opt/jax/inline.py:8 (glug)
                         # []
    c:i32[3] = add b 1   # /opt/jax/inline.py:5 (bar)
                         # []
  in (c,) : {  } }
```
XLA fuses the two additions into a single +3, as one might expect.
The resulting thunk metadata (from #8771) is:
```
Source locations: 
- /opt/jax/inline.py:17[<module>]
-- /opt/jax/inline.py:13[prog]
--- /opt/jax/inline.py:5[bar]

HLO: loop_add_fusion = s32[3]{0} fusion(Arg_0.1), kind=kLoop, calls=fused_add, metadata={op_name="jit(prog)/jit(main)/add" source_file="/opt/jax/inline.py" source_line=5}
Called HLO: 
fused_add {
  param_0 = s32[3]{0} parameter(0)
  constant_1 = s32[] constant(3)
  broadcast.6 = s32[3]{0} broadcast(constant_1), dimensions={}, metadata={op_name="jit(prog)/jit(main)/add" source_file="/opt/jax/inline.py" source_line=5}
  ROOT add.5 = s32[3]{0} add(param_0, broadcast.6), metadata={op_name="jit(prog)/jit(main)/add" source_file="/opt/jax/inline.py" source_line=5}
}
```
the point of this issue is that the thunk is basically doing an elementwise +3, but the metadata only records that this comes from line 5 of the input Python, which is
```python
return z + 1
```
which is not the full story.

This is, presumably, because an `HloInstruction` only has one `OpMetadata` struct associated to it, which in turn only has one `stack_frame_id` value - so the XLA representation doesn't have space to express that one addition can originate from multiple source lines.

This reduces the accuracy of tooling based on these metadata.

**Comment by gwenzek** [comment]
This seems to be tracked on Google internal bug tracker at b/311155137  

https://github.com/openxla/xla/blob/c1d5dfd2313d260421e135ec8d830e381bb182e3/xla/translate/mhlo_to_hlo/location_exporter.cc#L123-L125

Can this issue be ported to github if there is any activity  to follow up on it ?

**Comment by olupton** [comment]
> This seems to be tracked on Google internal bug tracker at b/311155137
> 
> https://github.com/openxla/xla/blob/c1d5dfd2313d260421e135ec8d830e381bb182e3/xla/translate/mhlo_to_hlo/location_exporter.cc#L123-L125
> 
> Can this issue be ported to github if there is any activity to follow up on it ?

I think that that is a little different?
One HLO instruction has one metadata structure associated with it.
Inside the metadata structure there is space for:
1. a single filename
2. a single line number
3. a list of stack frames

I believe the issue you link to is referring to the built-in profiler needing 1 and 2 because it doesn't support 3.

The issue I was raising here is different, namely that this:
> One HLO instruction has one metadata structure associated with it.

is itself a limitation, and capturing the metadata of the example I gave requires a single HLO instruction to have multiple metadata structures.
